common/browser.go

```diff
@@ -137,23 +136,29 @@ func (b *Browser) disposeContext(id cdp.BrowserContextID) error {
 		return fmt.Errorf("disposing browser context ID %s: %w", id, err)
 	}
 
-	b.contextsMu.Lock()
-	defer b.contextsMu.Unlock()
-	delete(b.contexts, id)
+	b.contextMu.Lock()
+	defer b.contextMu.Unlock()
+	b.context = nil
 
 	return nil
 }
 
-// getDefaultBrowserContextOrByID returns the BrowserContext for the given page ID.
+// getDefaultBrowserContextOrMatchedID returns the BrowserContext for the given browser context ID.
 // If the browser context is not found, the default BrowserContext is returned.
-func (b *Browser) getDefaultBrowserContextOrByID(id cdp.BrowserContextID) *BrowserContext {
-	b.contextsMu.RLock()
-	defer b.contextsMu.RUnlock()
-	browserCtx := b.defaultContext
-	if bctx, ok := b.contexts[id]; ok {
-		browserCtx = bctx
-	}
-	return browserCtx
+// If the existing browser context id doesn't match an error is returned.
+func (b *Browser) getDefaultBrowserContextOrMatchedID(id cdp.BrowserContextID) (*BrowserContext, error) {
+	b.contextMu.RLock()
+	defer b.contextMu.RUnlock()
+
+	if b.context == nil {
+		return b.defaultContext, nil
+	}
+
+	if b.context.id != id {
+		return nil, fmt.Errorf("missing browser context. Have: %s, want: %s", b.context.id, id)
```

Comment:
Good question, I'm not sure what the right way forward is. When i ran stress tests it didn't panic.

I prefer to be more defensive at this early stage of this change, but we're in a good place to be able to test this change in the next several weeks and if it panics we can investigate and switch to a warning.

WDYT?

Reply:
I agree with you both. By the way, could we make this error message slightly better (all lowercase and like a sentence) as follows? I believe the `getDefaultBrowserContextOrByID()` name was fine, but I'm fine with both.

```suggestion
		return nil, fmt.Errorf("missing browser context %s, current context is %s", id, b.context.id)
```

Reply:
> I prefer to be more defensive at this early stage of this change, but we're in a good place to be able to test this change in the next several weeks and if it panics we can investigate and switch to a warning. If we work with warn, then we might miss it and not get around to investigating the issue in time.
> 
> WDYT?

Ok, I'm ok leaving it as error at least by now :+1:

Reply:
It didn't take long to `panic` 🙂 So instead i'm switching back to working with the `defaultContext`. It occurs as soon as we try to connect multiple instances of k6 to a single browser.

My guess as to why a `browserContext` cannot be found when we receive a `onAttachedToTarget` event is because each of the k6 instances mirrors the `page`s from all other k6 instances, but we're not mirroring the creation of the `browserContexts`. It would be great if we didn't also have to create the `page` that is associated with a different k6 instance and not the current one 🤔 

I ran a quick experiment where i returned `nil` from `getDefaultBrowserContextOrMatchedID`:

```go
func (b *Browser) getDefaultBrowserContextOrMatchedID(id cdp.BrowserContextID) (*BrowserContext, error) {
	if b.context == nil {
		return nil, nil
	}

	if b.context.id != id {
		return nil, nil
	}

	return b.context, nil
}
```

and the caller would have to check for a nil `browserCtx`:

```go
	if browserCtx == nil {
		return // Ignore requests to attach target for other k6 instances
	}
```

Which actually all worked! So maybe there is a possibility of making this less complex and not needing to track `page`s from other k6 instances in the current k6 instance. I'll create an issue to look into this as it's not a priority for this current PR and issue that it is trying to resolve.

Reply:
> It would be great if we didn't also have to create the page that is associated with a different k6 instance and not the current one 🤔

That is how CDP works, but we can internally put a mechanism to manage that.

Reply:
> I'm not sure how returning nil is a better solution 😄 But I agree that making it straightforward to handle all these would be great.
Returning `nil` is just the hack, but essentially we wouldn't need to create a `page` if a `nil` `browserContext` was returned to the caller of `getDefaultBrowserContextOrMatchedID`, but i'm hoping that there are other things we can skip too, but yeah, this requires some deeper investigation.

Reply:
Reverted back to original logic: f6109efe8a0c752c6c13a2f6bb4e9e0d581db1ba

Reply:
To be honest, I don't think our previous implementation was correct anyway, but having https://github.com/grafana/xk6-browser/issues/933 open will address finding the proper solution for this. Resolving this discussion.